csp_billing_adapter/archive.py

```diff
@@ -18,24 +18,29 @@
 
 import functools
 import logging
+import json
 
 from csp_billing_adapter.config import Config
 from csp_billing_adapter.utils import retry_on_exception
 
 log = logging.getLogger('CSPBillingAdapter')
 
 DEFAULT_RETENTION_PERIOD = 6  # in months
+DEFAULT_BYTES_LIMIT = 0
```

Comment:
I think the "special" unlimited size case should be a negative number, suggest `-1`. A size limit of `-1` would generally be nonsensical and as such bring attention to the setting and it's special meaning. This should also be considered in combination with the `DEFAULT_RETENTION_PERIOD` where 0 is not special and it simply means "no records are retained". Similar as `0` is certainly a valid limit for memory limit setting `0` would indicate "no records are retained" and `0` has the same meaning for both settings. Having the same interpretation for a numerical value for related settings simplifies the thought requirement, IMHO.

Reply:
The smallest possible size of the archive is 2 bytes. Because a json string of an empty list is 2 bytes. So a value of 0 or 1 is non-sensical at least from a technical perspective. I can't think of examples off the top of my head but I'm certain I've used other tools where a value of 0 for a "limit" meant no limit. So I think there is precedence there. The affordances that I think we need are:

- Non-zero positive bytes limit
- Non-zero positive retention period limit
- No bytes limit (open discussion)
- Disable archive (retention period of zero)

I don't think we need two ways to disable the archive mechanism. If we want it then I would update the code to:

- Treat a 0 or 1 in bytes limit as "disable feature"
- Treat a 0 in retention period as "disable feature"
- Treat < 0 in bytes limit as no limit

This way min valid bytes limit is 2 which still makes no sense but prevents unnecessary handling to prevent infinite loop.

Reply:
Well actually from a memory perspective the smallest size is `53`
```
>>> sys.getsizeof(json.dumps('{}'))
53
```
and since `DEFAULT_BYTES_LIMIT` is supposed to set the memory usage limit for the retained records, driven by the k8s size limit of a config map we have to be concerned with the memory size.

It is fair to state that there is only 1 way to disable the feature and this is via the setting of the `DEFAULT_RETENTION_PERIOD`, then let's have a comment to that effect in the code.

My concern is not about the precedence that `0` has the special "no limit" meaning. My concern is that the reader of the code has 2 assignments in succession for which the value `0` has two distinct meanings with no explanation and where `0` is, based on naming of the variables, a valid setting that does not imply a special meaning. The question is how do we, in the code, at the point of assignment, make the special meaning evident? This can be done by using a value that on the surface makes not sense, `-1` for example, or it can be done via comment at the point assignment.

Ultimately we do not want to create a cognitive burden for the code reader that thinks of `0` as nothing special at this point and then ~90 lines later has to learn it is special after all.

Reply:
The measurement that matters is the size per character not size of the python object memory footprint.